HW-1 NAVER API로 영화 검색 앱 만들기

sangjinKIM93 opened the pull request:
영화이름을 입력하고 검색버튼을 누르면 하단에 리스트로 영화 목록 출력
목록에 있는 영화 아이템을 선택하면 웹브라우져로 이동
앞으로 스터디 할 내용들(MVP, MVVM, DataBinding 등등 사용하지 않고 과제 하기)

*사용한 라이브러리 - Retrofit2, Glide, Gson

Commit:
MainActivity.class , activity_main 이름 수정

각각 MovieListActivity.class, activity_movie_list로 수정

Commit:
actiity_movie_list.xml 레이아웃 작성

영화 검색 및 영화 목록 출력 레이아웃 작성
*recyclerView gradle 추가

Commit:
MoviewListViewHolder 클래스 생성

recyclerView 어답터에서 사용할 viewHolder 생성

Commit:
item_movie.xml 생성

리사이클러뷰 각 항목으로 들어갈 movie item 레이아웃 생성

Commit:
Movie data class 생성

Commit:
MovieListAdapter 생성

Commit:
recyclerView Id 추가

Commit:
MovieListActivity에서 recyclerView 초기화

Commit:
use-permission INTERNET 추가

Commit:
Retrofit2 gradle 추가

추가적으로 gson, gson-converter도 추가

Commit:
RetrofitService Interface 추가

keyword로 영화를 검색할 수 있는 requestMovieList 메서드

Commit:
retrofit & retrofitService 객체 생성

Commit:
ResponseData class 추가

받는 데이터 json 구조가 array 형태가 아니라 jsonObject로 한번 감싼 구조라
이를 풀기 위해서 ResponseData class 추가.

Commit:
MovieListActivity에 getMoveList() 추가

retrofitService를 통해서 json 형식의 movie 데이터 받기

Commit:
add refreshList()

기존 onCreate에서 초기화 하던 recyclerView 셋팅을 메서드로 분리
*추가적으로 movieList type을 array -> list 변경

Commit:
retrofit response에 따른 결과 출력 메서드

데이터가 있을 경우 리사이클러뷰 최신화
데이터가 업을 경우 "자료가 없다"는 텍스트뷰 출력

Commit:
검색 버튼 클릭 이벤트 추가

클릭시 레트로핏을 통해서 자료 받아오기

Commit:
add deleteHTMLTag()

json으로 받아온 제목에 <b></b>가 붙는 현상이 있어 이를 제거해주는 메서드를 생성 및 onBindViewHolder에서 적용

Commit:
add Glide

글라이드를 통한 이미지 출력.
*adapter에서 글라이드 사용을 위해서 매개변수로 context 추가
*deleteHTMLTag에서 변수를 잘못지정하여 수정

Commit:
add MovieDetailActivity & activity_movie_detail.xml

Commit:
add onItemClickListener

리사이클러뷰 각 항목 클릭 이벤트를 처리하기 위해서 onItemClickListener를 생성
*각 포지션 값 전달을 위한 getItemId() 생성

Commit:
intent 생성

해당 영화의 link를 intent에 담아서 MovieDetailActivity로 전달

Commit:
webView 생성 및 Link 받아서 출력

*URL_CLEARTEXT 오류로 인해서 매니페스트에 설정 추가

Commit:
ImageView와 textView의 높이를 비율로 수정

기존 고정 dp로 높이를 설정한 것을 비율로 변경함.
*layout_constraintDimensionRatio를 사용하기 위해서 부모 layout을 constraint로 변경

Commit:
검색결과 없을시 nullCheck 방법 수정

기존 방법은 empty인 경우에 대한 분기가 되지 않아 isNorOrEmpty로 수정

Commit:
string 직접 입력을 string 리소스로 관리

RyuC0605 commented:
안녕하세요. 류승윤 입니다. 
상진님 반갑습니다. 
석준님께서 말씀 하신대로 커밋을 엄청 쪼개서 하셨네요. 너무 과할정도로;;;
일단 코드 리뷰에 앞써서 코드를 빌드해서 한번 돌려 봤습니다. 

앱이 요구사항과 맞지 않는 부분이 보이더군요.
가장 크리티컬한 부분은 영화를 검색 후 목록을 클릭하면 앱이 죽더군요. 

일딴 원인은 찾았고, 제가 알려드리면 실력 향상에 도움이 되지 않을 것 같아 
남기지 않으려 했으나, 첫 시간이니 간단히 말씀드리겠습니다. 
다음부턴 힌트만 드리겠습니다.

01. 앱 죽는 현상 : 인텐트에 값을 넘길때 넘기는 쪽은 LINK 받는쪽은 Link 로 해두셨더군요. 
대소문자를 구분하기 때문에 이런 경우 상수로 설정하는게 안전합니다.

02. 요구사항 준수 : 요구사항은 webView 서브화면 제작이 아닌, 외부 브라우져 호출입니다.

세부사항은 각 커밋에 해당하는 부분에 코멘트를 남기겠습니다. 

연휴 잘 보내세요.

RyuC0605 commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/MovieListAdapter.kt:
```diff
@@ -14,9 +14,17 @@ class MovieListAdapter (private val movieList: List<Movie>) : RecyclerView.Adapt
+        var strEdited = str.replace("<b>", "")
+        strEdited = str.replace("</b>", "")
+
+        return strEdited
```
<b> 태그를 수동으로 지워주는 것도 방법이지만, 컬러나 기타 각종 모든 태그를 이렇게 처리하려면 일이 커지죠.
이럴땐 Html.fromHtml 을 추천드립니다. 검색창에 검색해 보세요.

sangjinKIM93 replied:
Html.fromHtml 라는 것을 새롭게 알게되었네요

반영했습니다. 감사합니다:)

RyuC0605 commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/MovieListActivity.kt:
```diff
@@ -42,5 +52,15 @@ class MovieListActivity : AppCompatActivity() {
+    private fun noResultMessage(){
+        movieListView.visibility = View.INVISIBLE
+        guideMessageToInput.visibility = View.VISIBLE
+
+        guideMessageToInput.text = "검색 결과가 없습니다."
```
"검색 결과가 없습니다." 를 string.xml 로 리소스로 관리하는 것이  추후 다국어 라든지 여러모로 좋습니다.

sangjinKIM93 replied:
kotlin 코드상의 String을 처리하지 않았네요. 반영했습니다!

RyuC0605 commented on 2005/sangjinKIM93/newproject/app/src/main/res/layout/activity_movie_list.xml:
```diff
@@ -38,6 +38,20 @@
+        android:id="@+id/guideMessageToInput"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:text="영화를 검색해주세요."
```
역시 하드코딩은 좋지 않습니다. 위에 언급한 string.xml 로 빼는걸 추천드립니다.

sangjinKIM93 replied:
추후 수정 commit이 있는데 이전 commit이 먼저 남아있군요. 수정했습니다!

RyuC0605 reviewed (changes requested):
01. 앱 죽는 현상 : 인텐트에 값을 넘길때 넘기는 쪽은 LINK 받는쪽은 Link 로 해두셨더군요. 
대소문자를 구분하기 때문에 이런 경우 상수로 설정하는게 안전합니다.

02. 요구사항 준수 : 요구사항은 webView 서브화면 제작이 아닌, 외부 브라우져 호출입니다.

세부사항은 각 커밋에 해당하는 부분에 코멘트를 남기겠습니다. 

연휴 잘 보내세요.

Commit:
webView 삭제 및 ACTION_VIEW로 변경

Commit:
html tag 삭제 방식 변경 (Html.fromHtml)

Commit:
string 하드코딩 수정

Commit:
Naver API 키 숨기기

sangjinKIM93 commented:
크리티컬 한 리뷰에 대하여

1. 다시 한번 상수의 중요성을 느낍니다. 이번 기회에 습관을 들여야겠네요.
2. 요구사항을 잘못 이해한 부분 수정했습니다.(webView -> ACTION_VIEW)

commit 단위는 세부적으로 쪼개려다 보니까 오히려 읽는 사람이 읽기 부담스러운 커밋이 되었네요. 추후 커밋에서 좀 더 신경써보겠습니다.

값진 피드백 정말 감사드리고 승윤님도 즐거운 연휴 보내세요 :)

RyuC0605 commented:
approval 했었는데 상단 새로고침(?) 비슷하게 생긴 아이콘 건드렸다가 뭔가 꼬였네요;;;

RyuC0605 commented:
approval  에서  was requested for review 로 바꼈는데 다시 approval  시키는 방법을 못찾고 있네요.

RyuC0605 commented:
상단의 Reviewers 에 제 아이디 옆 새로고침(?) 비슷한 아이콘 건드렸다가 승인이 풀려버렸었네요.
지금은 다시 approval 했습니다. 휴... 식겁하게 하네요 깃허브...

Commit:
검색 기능 관련 keyboard셋팅

1. ImeOption 설정
2. 키보드 보이기 / 숨기기 설정
3. btn text를 기존 '완료'->'검색'으로 수정

Commit:
CardView 레이아웃 수정 및 default Image 추가

Commit:
default Img 파일 추가

sangjinKIM93 commented:
깃헙.... 까다로운 친구 같습니다 ㅎㅎ

승윤님 코드랑 다른분들 리뷰 보면서 부족한 부분 추가 수정하였습니다.

Commit:
movieList 최신화 방법 변경 & 안내 문구 textview 삭제

1. 기존 리사이클러뷰를 모두 다시 셋팅하는 방법에서 notifyDataSetChanged()를 통한 리스트 최신화로 변경
2. 기존 검색 결과 안내 textview를 지우고 toast로 경고

sangjinKIM93 commented:
깃헙.... 까다로운 친구 같습니다 ㅎㅎ

승윤님 코드랑 다른분들 리뷰 보면서 부족한 부분 추가 수정하였습니다.

sjjeong commented:
![image](https://user-images.githubusercontent.com/4777586/80919649-1f73ed00-8da6-11ea-895d-6504f6f045d0.png)
master로 보내면 안되고 2005/sangjinKIM93으로 보내야 합니다.
일단 제가 수정했는데 앞으로는 pr만드실 때 주의해주세요

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+import com.sangjin.newproject.R
+import kotlinx.android.synthetic.main.item_movie.view.*
+
+class MovieListAdapter (private val movieList: ArrayList<Movie>, private val context: Context) : RecyclerView.Adapter<MovieListViewHolder>(){
```
생성자로 Context를 받는 부분 제거해주시고
List는 Adapter외부와 독립적으로 내부에 별도 객체로 가지고 있게 해주세요

sangjinKIM93 replied:
2d64251
수정했습니다.

질문이 있습니다!
1. private val movieList: ArrayList<Movie>를 위처럼 매개변수로 받아 바로 생성하는 것은 내부에 객체를 별도로 만드는 것과 어떻게 다른가요?
2. 내부에 별도 객체로 만들라고 하셨는데, 그렇게 하지 않을 경우 어떤 문제점이 있나요?

sjjeong replied:
지금 변경하신 내용도 객체는 같게 됩니다.
Activity가 가지고 있는 list의 hash와 Adapter가 가지고 있는 list의 hash를 출력해보세요

객체가 같으면 Activity에서 내용을 수정하면 Adapter에 있는 내용도 같이 변경되기 때문에 버그가 발생할 수 있습니다.
그래서 별도 객체를 가지도록 하는게 좋습니다.

sangjinKIM93 replied:
0c47078
이해되었습니다. 같은 메모리 주소를 참조하고 있으면 같은 영향을 받으니 문제가 생길 수 있겠군요. 
adapter에 새로운 arrayList 객체를 생성하고 addList()로 최신화 해주는 것으로 수정하였습니다.
좋은 지적 감사합니다:)

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+        val item = LayoutInflater.from(parent.context).inflate(R.layout.item_movie, parent, false)
+
+        item.setOnClickListener{
+            var position: Long = item.tag as Long
```
tag를 사용하는 것은 좋은 방법이 아닙니다.
position은 ViewHolder에 있는 adapterPosition으로 가져오면 됩니다.

sangjinKIM93 replied:
e828999
수정했습니다.

찾아보니 불필요한 코드였네요. 감사합니다.

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    override fun onBindViewHolder(holder: MovieListViewHolder, position: Int) {
+        holder.containerView.movieTitleTV.text = movieList.get(position).title.htmlToString()
+
+        Glide.with(context)
+            .load(movieList.get(position).image)
```
movieList.get(position)가 중복으로 사용되는데 val로 빼서 사용해주세요

sangjinKIM93 replied:
2d7c884
수정했습니다.

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    fun String.htmlToString() : String{
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            return Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY).toString()
+        } else {
+            return Html.fromHtml(this).toString()
+        }
```
아래와 같이 변경할 수 있습니다.
```kotlin
return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
    Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY).toString()
} else {
    Html.fromHtml(this).toString()
}
```

sangjinKIM93 replied:
704ce23
수정했습니다.

새롭게 알게 된 부분이네요. 감사합니다:)

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/adapter/Movie.kt:
```diff
@@ -0,0 +1,16 @@
+)
+
+data class Movie(
+    var title: String = "",
+    var link: String = "",
+    var image: String = "",
+    var subtitle: String = "",
+    var pubDate: String = "",
+    var director: String = "",
+    var actor: String = "",
+    var userRating: String = ""
```
val로 변경해주시고 default를 지정하는 것은 제거해주세요

sangjinKIM93 replied:
03213b4
수정했습니다!

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/MovieListActivity.kt:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * 키패드 보여주기
+     */
+    fun showKeyPad(){
```
show/hide keypad는 private으로 변경해주세요

sangjinKIM93 replied:
0fab89b
수정했습니다!

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/MovieListActivity.kt:
```diff
@@ -0,0 +1,130 @@
+        if(TextUtils.isEmpty(keyWord)){
+            Toast.makeText(this, R.string.no_keyword, Toast.LENGTH_LONG).show()
+        }
+        else{
```
if의 }와 같은 라인에 있게 해주세요

sangjinKIM93 replied:
f6f1bbe
수정했습니다!

sjjeong commented on 2005/sangjinKIM93/newproject/app/src/main/java/com/sangjin/newproject/MovieListActivity.kt:
```diff
@@ -0,0 +1,130 @@
+        movieListView.adapter = movieListAdapter
+
+        //각 항목 클릭시 이벤트
+        movieListAdapter.onItemClickListener = {position ->
+            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(movieList.get(position).link))
+            startActivity(intent)
+        }
```
생성자로 넘기면 nullable 처리 없이 nonnull로 사용할 수 있습니다.

sangjinKIM93 replied:
말씀하신 의도를 잘 이해하지 못 했습니다. 
movieList.get(position).link 를 let을 활용해서 보내라는 말씀이신가요?

sjjeong replied:
현재 MoviewListAdapterdk안에 onItemClickListener을 말하는 거였습니다.

sangjinKIM93 replied:
a6db658
수정은 했는데 코드 진행방식이 적절한지 잘 모르겠네요 확인부탁드립니다!

onItemClickListener를 생성자로 넘길 생각은 못 해봤는데 새롭게 알고 갑니다.
감사합니다:)

Commit:
MovieListAdapter 매개변수로 받는 context, list 수정

context는 onCreateViewHolder에서 생성하는 것으로 변경
list는 내부에 별도 객체로 생성

Commit:
position 받는 방법 변경

tag -> viewholder.adapterPosition

Commit:
moviewList.get(Item) 중복 문제 변수로 만들어 해결

Commit:
return if 활용으로 불필요한 return 구문 제거

Commit:
data class 변수 var을 val로 변경 , default 제거

Commit:
hide/show keypad private로 변경

Commit:
if else 줄 간격 조정

Commit:
MovieListAdapter에 addList()메서드 추가

기존 매개변수로 list를 받는 방법은 ativity와 adapter가 같은 객체를 참조해서 연쇄 작용으로 인한 문제가 발생할 수 있어. adapter에서 별도의 arrayList 객체를 생성하고 메서드를 통해서 이를 최신화 하는 방법으로 문제를 해결

Commit:
onItemClickListener 생성자로 넘기기

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
